File: automations/seo-optimizer/02-agents/_shared/llm_client.py
"""LLM client wrapper around Anthropic SDK (via OpenRouter).

Uses prompt caching for system prompts (Anthropic feature, supported by
OpenRouter passthrough). This is critical for cost — analyst + writer
share large system prompts that we want cached.

The model is chosen via env SEO_OPTIMIZER_MODEL (default:
anthropic/claude-sonnet-4.5).

Usage:

    from _shared.llm_client import generate, generate_json

    text = generate(
        system="You are a senior SEO consultant.",
        user="Analyze this article: ...",
        cache_system=True,
        max_tokens=2000,
    )

    parsed = generate_json(
        system="...",
        user="...",
        schema={"type": "object", ...},
    )
"""
from __future__ import annotations

import json
import os
from typing import Any

import structlog
from anthropic import Anthropic
from anthropic.types import TextBlock
from tenacity import (
    retry, retry_if_exception_type, stop_after_attempt, wait_exponential,
)


log = structlog.get_logger()

DEFAULT_MODEL = "anthropic/claude-sonnet-4.5"


def _client() -> Anthropic:
    """Anthropic client configured for OpenRouter passthrough."""
    if hasattr(_client, "_cached"):
        return _client._cached  # type: ignore[attr-defined]

    api_key = os.environ.get("OPENROUTER_API_KEY")
    if not api_key:
        raise RuntimeError("OPENROUTER_API_KEY not set")

    # OpenRouter is OpenAI-compatible AND Anthropic-compatible. For prompt
    # caching to actually work we use the Anthropic SDK pointing to OpenRouter's
    # /anthropic/ proxy endpoint.
    client = Anthropic(
        api_key=api_key,
        base_url="https://openrouter.ai/api/v1",
    )
    _client._cached = client  # type: ignore[attr-defined]
    return client


def _model() -> str:
    return os.environ.get("SEO_OPTIMIZER_MODEL", DEFAULT_MODEL)


@retry(
    retry=retry_if_exception_type((Exception,)),
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=2, min=2, max=16),
    reraise=True,
)
def _call_llm(
    *,
    system: str | list[dict],
    user: str,
    max_tokens: int,
    temperature: float,
) -> tuple[str, dict]:
    """Low-level call; returns (text, usage_dict)."""
    client = _client()

    # Build system blocks (with cache_control if requested via list form)
    if isinstance(system, str):
        system_blocks = [{"type": "text", "text": system}]
    else:
        system_blocks = system

    resp = client.messages.create(
        model=_model(),
        max_tokens=max_tokens,
        temperature=temperature,
        system=system_blocks,
        messages=[{"role": "user", "content": user}],
    )

    text = "".join(
        block.text for block in resp.content if isinstance(block, TextBlock)
    )
    usage = {
        "input_tokens": getattr(resp.usage, "input_tokens", 0),
        "output_tokens": getattr(resp.usage, "output_tokens", 0),
        "cache_creation_input_tokens": getattr(resp.usage, "cache_creation_input_tokens", 0),
        "cache_read_input_tokens": getattr(resp.usage, "cache_read_input_tokens", 0),
    }
    return text, usage


def generate(
    *,
    system: str,
    user: str,
    max_tokens: int = 2000,
    temperature: float = 0.3,
    cache_system: bool = True,
) -> str:
    """Generate text. Returns just the text."""
    if cache_system:
        system_arg: str | list[dict] = [
            {"type": "text", "text": system, "cache_control": {"type": "ephemeral"}},
        ]
    else:
        system_arg = system

    text, usage = _call_llm(
        system=system_arg, user=user, max_tokens=max_tokens, temperature=temperature,
    )
    log.info("llm_generate", usage=usage, output_chars=len(text))
    return text


def generate_json(
    *,
    system: str,
    user: str,
    max_tokens: int = 4000,
    temperature: float = 0.1,
    cache_system: bool = True,
) -> dict[str, Any]:
    """Generate JSON output. Parses and returns dict.

    Robust to LLM wrapping JSON in code fences. Raises ValueError if no parseable
    JSON found.
    """
    text = generate(
        system=system,
        user=user + "\n\nRespond with ONLY a valid JSON object. No code fences, no commentary.",
        max_tokens=max_tokens,
        temperature=temperature,
        cache_system=cache_system,
    )
    # Strip code fences if present
    cleaned = text.strip()
    if cleaned.startswith("```"):
        # Drop first and last fence lines
        lines = cleaned.splitlines()
        cleaned = "\n".join(lines[1:-1]) if len(lines) >= 2 else cleaned

    try:
        return json.loads(cleaned)
    except json.JSONDecodeError as exc:
        log.error("llm_json_parse_failed", text=text[:500], error=str(exc))
        # Try to extract first {...} block
        import re
        m = re.search(r"\{.*\}", cleaned, re.DOTALL)
        if m:
            try:
                return json.loads(m.group(0))
            except json.JSONDecodeError:
                pass
        raise ValueError(f"LLM returned non-JSON: {text[:200]}") from exc


def call_with_usage(
    *,
    system: str,
    user: str,
    max_tokens: int = 2000,
    temperature: float = 0.3,
    cache_system: bool = True,
) -> tuple[str, dict]:
    """Like generate() but also returns the usage dict.

    Used by /writer to record tokens_input/output in opportunity_rewrites.
    """
    if cache_system:
        system_arg: str | list[dict] = [
            {"type": "text", "text": system, "cache_control": {"type": "ephemeral"}},
        ]
    else:
        system_arg = system

    text, usage = _call_llm(
        system=system_arg, user=user, max_tokens=max_tokens, temperature=temperature,
    )
    return text, usage

File: automations/seo-optimizer/supabase/functions/_shared/llm-client.ts
// _shared/llm-client.ts
// Wrapper sobre Anthropic SDK (via OpenRouter passthrough).
// Soporta prompt caching del system prompt para bajar costo.

import Anthropic from "npm:@anthropic-ai/sdk@0.32.1";

const DEFAULT_MODEL = "anthropic/claude-sonnet-4.5";

let _client: Anthropic | null = null;

function getClient(): Anthropic {
  if (_client) return _client;
  const apiKey = Deno.env.get("OPENROUTER_API_KEY");
  if (!apiKey) throw new Error("OPENROUTER_API_KEY not set");
  _client = new Anthropic({
    apiKey,
    baseURL: "https://openrouter.ai/api/v1",
  });
  return _client;
}

function getModel(): string {
  return Deno.env.get("SEO_OPTIMIZER_MODEL") ?? DEFAULT_MODEL;
}

export interface LlmUsage {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens?: number;
  cache_read_input_tokens?: number;
}

export interface CallWithUsageArgs {
  system: string;
  user: string;
  maxTokens?: number;
  temperature?: number;
  cacheSystem?: boolean;
}

export async function callWithUsage(args: CallWithUsageArgs): Promise<{ text: string; usage: LlmUsage }> {
  const client = getClient();
  const cacheSystem = args.cacheSystem ?? true;
  const systemBlocks = cacheSystem
    ? [{ type: "text" as const, text: args.system, cache_control: { type: "ephemeral" as const } }]
    : [{ type: "text" as const, text: args.system }];

  // deno-lint-ignore no-explicit-any
  const resp: any = await client.messages.create({
    model: getModel(),
    max_tokens: args.maxTokens ?? 2000,
    temperature: args.temperature ?? 0.3,
    system: systemBlocks,
    messages: [{ role: "user", content: args.user }],
  });

  // deno-lint-ignore no-explicit-any
  const text = (resp.content as any[])
    .filter((b: { type: string }) => b.type === "text")
    .map((b: { text: string }) => b.text)
    .join("");

  const usage: LlmUsage = {
    input_tokens: resp.usage?.input_tokens ?? 0,
    output_tokens: resp.usage?.output_tokens ?? 0,
    cache_creation_input_tokens: resp.usage?.cache_creation_input_tokens,
    cache_read_input_tokens: resp.usage?.cache_read_input_tokens,
  };
  return { text, usage };
}

export async function generate(args: CallWithUsageArgs): Promise<string> {
  const { text } = await callWithUsage(args);
  return text;
}

export async function generateJson<T = Record<string, unknown>>(args: CallWithUsageArgs): Promise<T> {
  const text = await generate({
    ...args,
    user: args.user + "\n\nRespond with ONLY a valid JSON object. No code fences, no commentary.",
    maxTokens: args.maxTokens ?? 4000,
    temperature: args.temperature ?? 0.1,
  });
  return parseLlmJson<T>(text);
}

export function parseLlmJson<T = Record<string, unknown>>(text: string): T {
  let cleaned = text.trim();
  if (cleaned.startsWith("```")) {
    const lines = cleaned.split("\n");
    cleaned = lines.slice(1, -1).join("\n");
  }
  try {
    return JSON.parse(cleaned) as T;
  } catch {
    const match = cleaned.match(/\{[\s\S]*\}/);
    if (!match) throw new Error(`LLM returned non-JSON: ${text.slice(0, 200)}`);
    return JSON.parse(match[0]) as T;
  }
}

export function getModelName(): string {
  return getModel();
}
